src/dhtproto/client/legacy/internal/helper/Handshake.d

```diff
@@ -0,0 +1,211 @@
+
+            Boolean flag that will be set to `true` once all DHT nodes have
+            been reached by the handshake
+
+        ***********************************************************************/
+
+        private bool all_dht_nodes_connected = false;
+
+        /***********************************************************************
+
+            Delegate to be called by the handshake once all DHT nodes have
+            been reached
+
+        ***********************************************************************/
+
+        private void dg ()
+        {
+            this.all_dht_nodes_connected = true;
+            this.outer.log.info("All DHT nodes connected!");
+            if (this.suspended()) this.resume();
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            if (!this.all_dht_nodes_connected) this.suspend();
+        }
+    }
+
+
+    /***************************************************************************
+
+        Task that suspends until at least one node is connected
+
+    ***************************************************************************/
+
+    private class OneDhtNodeConnected : Task
+    {
+        import swarm.Const : NodeItem;
+
+        /***********************************************************************
+
+            Boolean flag that will be set to `true` once at least one DHT node
+            has been reached by the handshake
+
+        ***********************************************************************/
+
+        private bool at_least_one_node_connected = false;
+
+        /***********************************************************************
+
+            Delegate to be called by the handshake each time an individual
+            node connects
+
+        ***********************************************************************/
+
+        private void dg (NodeItem item)
+        {
+            if (!this.at_least_one_node_connected)
+            {
+                this.at_least_one_node_connected = true;
+                if (this.suspended()) this.resume();
+            }
+
+            this.outer.log.trace("Connected to DHT node {}:{}",
+                                 item.Address, item.Port);
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            if (!this.at_least_one_node_connected) this.suspend();
+        }
+    }
+
+
+    /***************************************************************************
+
+        Logger for the handshake
+
+    ***************************************************************************/
+
+    private Logger log;
```

Comment:
Optional style point: we tend to prefer `///` comments for everything expect functions, these days.